### PR TITLE
emoji completion based on the emo::ji_completion function.

### PR DESCRIPTION
**Pull request by romainfrancois**
If the emo :package: is not available, then we just get empty completions.

This PR replaces #1459

### CODE REVIEW

**Comment by romainfrancois** [comment]
this gives completions based on the regex "[:][a-zA-Z0-9_]+$" in some places
 - comments (incl roxygen comments, bc if the token matches, then typical roxygen completion make no sense
 - strings (single and multi line). I believe this is alright because things like `:ca` won't match file names. 
 - markdown text
 
Later I'd like to also support emoji completions in the commit messages, but this is another PR I guess because there is currently unless I missed it scaffolding for completions in there. 

@kevinushey I believe this is ready for review
@jmcphers thanks for the comments on #1459 this definitely motivated me to submit this much cleaner PR.

**Comment by romainfrancois** [comment]
one thing to consider though is that ace support for emoji does not look ideal: 
https://github.com/ajaxorg/ace/issues/460

i.e. when you insert 0️⃣ and then try to delete it, you have to delete 3 times, i.e. it does not understand the varying width. 

This is not a problem of this PR, but we might get some complains because this PR makes it really easy to insert emojis and then witness the ace bug.

File: src/cpp/session/modules/SessionRCompletions.R [hunk 1]
```diff
@@ -2469,6 +2475,22 @@ assign(x = ".rs.acCompletionTypes",
    )
 })
 
+.rs.addFunction("getCompletionsEmoji", function(token){
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: style: prefer space, as in `function(token) {`

File: src/cpp/session/modules/SessionRCompletions.R [hunk 2]
```diff
@@ -2469,6 +2475,22 @@ assign(x = ".rs.acCompletionTypes",
...
+   tryCatch({
+      results <- emo::ji_completion( sub( "^[:]", "", token ) )
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: style: avoid spaces around function arguments

File: src/cpp/session/modules/SessionRCompletions.R [hunk 3]
```diff
@@ -2469,6 +2475,22 @@ assign(x = ".rs.acCompletionTypes",
...
+      .rs.makeCompletions(
+         token = token, 
+         results = unname(results),
+         packages = names(results), # this is not a pkg, but we use this to store the matched emoji aliases
+         quote = FALSE, 
+         type = .rs.acCompletionTypes$EMOJI, 
+         excludeOtherCompletions = TRUE
+      )
+   }, error = function(e){
```
Comment (kevinushey) [inline src/cpp/session/modules/SessionRCompletions.R]: style: `function(e) {`

File: src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java [hunk 4]
```diff
@@ -174,6 +175,13 @@ private CompletionResult narrow(final String token,
             if (StringUtil.isSubsequence(basename(qname.name), tokenFuzzy, true))
                newCompletions.add(qname);
          }
+         else if(qname.type == RCompletionType.EMOJI)
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java]: space here: `if (qname.type ...`

File: src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java [hunk 5]
```diff
@@ -174,6 +175,13 @@ private CompletionResult narrow(final String token,
...
+         {
+            // narrowing based on the .source
+            if( qname.source.contains(tokenEmoji) ){
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java]: style?

File: src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java [hunk 6]
```diff
@@ -865,6 +873,15 @@ private void doAddDisplayNameGeneric(SafeHtmlBuilder sb)
                RES.styles().completion(),
                name);
 
+         // Display the alias, e.g. :cat: for the emoji completions
+         if( type == RCompletionType.EMOJI ){
```
Comment (kevinushey) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/console/shell/assist/CompletionRequester.java]: style?

**Comment by kevinushey** [comment]
A couple of minor issues I noticed while testing this PR:

1. Attempting to insert a completion with a plain `:` doesn't work; e.g. `":<TAB>`.
2. Emoji completions show up when using e.g. `':<TAB>` but not with `":<TAB>`; ie, the type of quotation character seems to affect behavior here.

You are indeed right that Ace's support for astral plane emoji is not great, which makes me a bit wary of taking this PR right now. (Note that in addition to requiring multiple 'delete' keypresses to remove an emoji, the cursor position is also drawn in the incorrect position on lines containing emoticons that don't match the monospaced font's width).

We've also even had reports that some users with documents containing multibyte emojis with e.g. 3 or more characters can cause document corruption when inserted / deleted in a document that is subsequently saved so there is definitely a lot of hairiness here!

Unfortunately, Ace support for emojis is overall not great and we're somewhat limited based on what it can handle. :-/

(Try copy + pasting 👩‍❤️‍💋‍👩 into RStudio for some extra weirdness...)

**Comment by romainfrancois** [comment]
I think the style issues are fixed. although : 

capture d ecran 2017-08-23 a 09 13 03

and: 

capture d ecran 2017-08-23 a 09 17 37

but these files are not relevant to this PR, so I'll let you handle them.

**Comment by romainfrancois** [comment]
About the issues related to ace, I understand your concern, but I might argue that this is orthogonal to this PR. 

right now 👨‍💻 go to some other place where they can type emoji, or use `emo::ji` on the console and then copy and paste it. so it's harder to get emojis, but once you get them you still witness the same ace issues. 

oth✋ maybe having it super easy to insert emojis and to therefore have these ace issues appearing will annoy people and cause some extra burden. But maybe having legions of 😠 R users might have an 👍 impact on https://github.com/ajaxorg/ace/issues/460 

It's odd indeed about the :kiss_woman_woman: emoji, but it's a complex one: 

capture d ecran 2017-08-23 a 09 51 48

There's definitely some issue with the ZWJ getting somewhat converted to a CR: 

capture d ecran 2017-08-23 a 10 02 16

**Comment by romainfrancois** [comment]
🙀 

capture d ecran 2017-08-23 a 10 13 45

**Comment by jmcphers** [comment]
> we might get some complains because this PR makes it really easy to insert emojis and then witness the ace bug

Yes, unfortunately I think we're going to need to wait for Ace to handle emojis properly before we start adding affordances for inserting them. Otherwise as you note a lot of users are going to be frustrated that their newly emoji-laden files suddenly have baffling cursor behavior and unpredictable text manipulation. 

I'm going to close this for now since we don't have a clear path forward on the Ace bits, but my hope is that they'll fix this upstream by the time we take another drop of the editor, and that we can reopen this for consideration at that time.

**Pull request closed without merging**